### PR TITLE
Separate Z speeds from XY speeds and reduce global variables, wrong home button in TFT70, long file name display during printing from onboardSD & other bugfixes

### LINKED ISSUE #1485: Z motion bug
### Description

Z feed rate is the same as X and Y, meaning that if you use the manual buttons to Move Z it's too fast and the stepper can not move the Z this fast. This is for Slow, Normal, and Fast. I have changed the Slow speed to make it possible to use the Move Z function but that also means that X and Y are now at a crawl.

### Steps to reproduce

1. Press Menu button
2. Press Move button
3. Press Z up or Z down button(s)

**Expected behavior**

Z should raise and lower at Slow, Normal and Fast speeds (but they are all too high)

**Actual behavior**

Grinding Sound and no motion

### Hardware Variant

BIGTREETECH TFT35-E3 V3.0

### TFT Firmware Version & Main Board Firmware details

BIGTREETECH SKR Mini E3 V2

### PR DESCRIPTION
**Pull request by guruathwal**
- Separate Z speeds from XY speeds (#1485 ).
- Reduce global variables.
- Fix long file name display during printing from onboard SD #1245. (This uses 4 bytes of RAM  in comparison to #1488 which uses 256 bytes of RAM).
- Fix Babystep menu opening on pressing the home button after a print is finished on TFT70.
- Removed scroll message from print summary and display "click for Summary" message. 
- Fix print summary message in popup window overflow on devices with 480x272 displays. 
- Fix double value in Numpad when values are larger than 1000.
- Add light-weight string to double function without exponential support
- Remove duplicate code.

Fixes #1485

### CODE REVIEW

**Comment by digant73** [comment]
Isn‘t it more safe to always initialize nextUpdateTime on each menu where it is used? Just thinking to the scenario where you move from a menu with very high delay to one with low delay

**Comment by kisslorand** [comment]
It was like this (one variable for nextWhateverTime) before and I changed it so one timer for a function/query would not interfere with another. Let's say you update the nextWhatever for coordinateQuery but it will interfere with the Toggle or speedQuery. nextUpdateTime is changed in various places with various delay times. That's nonsense.
You can get into a situation when something will never be refreshed because other function is constantly updating the nextUpdateTime.

File: TFT/src/User/Menu/Move.c [hunk 1]
```diff
@@ -57,6 +57,8 @@ void storeMoveCmd(AXIS xyz, int8_t direction)
 void menuMove(void)
 {
   KEY_VALUES key_num;
+  COORDINATE lastPos;
+  COORDINATE curPos;
```
Comment (kisslorand) [inline TFT/src/User/Menu/Move.c:62]: lastPos, curPos are never used

File: TFT/src/User/Menu/common.c [hunk 2]
```diff
@@ -82,3 +82,18 @@ const ITEM itemExtLenSteps[ITEM_EXT_LEN_NUM] =
...
+  if (curTime > (lastTime + duration))
+  {
+    lastTime = curTime;
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
```
Comment (kisslorand) [inline TFT/src/User/Menu/common.c:99]: A lower duration will reset lastTime so nextScreenUpdate() will always be false for higher duration query.

**Comment by guruathwal** [comment]
@digant73  Changing the logic to store last check time should take care of that.

@kisslorand 
> Let's say you update the nextWhatever for coordinateQuery but it will interfere with the Toggle or speedQuery. nextUpdateTime is changed in various places with various delay times. That's nonsense.
> You can get into a situation when something will never be refreshed because other function is constantly updating the nextUpdateTime.

The time update is changed for menu loop only and nothing is changed for global loops. There are no two menu functions running at the same time. All menu functions are destroyed as soon as they are exited. Do you think it is possible that any function that is not being called in the active menu loop will run by itself? 
`coordinateQuery` and `SpeedQuery` do not have any watchdog timers. They are called only when needed and along with display toggles by those menu loops which need fresh data and the timer is managed by those menu loops. Even `probeHeightQueryCoord` is also run only when a menu(mesh editor) needs it to run. Still, I have reverted it.

> Every parameter, that needed periodic update had its own watchdog with a predefined update period. Now, with the unified nextUpdateTime several parameters have the same watchdog with an ambiguous, not predictable update period.

There is no change in global loops which require separate timers. This change is in screen updates only.

**Comment by kisslorand** [comment]
@guruathwal 
Since my comment you've made changes so the things I wrote are not a concern anymore. Hope noone will misuse `nextScreenUpdate()` in the future.

**Comment by kisslorand** [comment]
> * Fix long file name display during printing from onboardSD #1245. (This uses 4 bytes of RAM  in comparison to #1488 which uses 256 bytes of RAM).

PR #1488 has been rewritten. I finally find a way to not disturb the filename when printing from an outside USB host that uses the onboard SD while returning to my original idea to use file index, which indeed takes only 4 bytes extra from RAM.

I will upload a short video to show why my original idea gave me headaches, I had the same things coded as you did.

**Comment by kisslorand** [comment]
> * Fix long file name display during printing from onboardSD #1245. (This uses 4 bytes of RAM  in comparison to #1488 which uses 256 bytes of RAM).

How about 0 (ZERO) bytes of RAM? Check PR #1488, the one that started the long filename issue fix.

**Comment by guruathwal** [comment]
> How about using 0 (ZERO) bytes of extra RAM? Check PR #1488.

To what did you compare #1488's RAM usage to??

**Comment by kisslorand** [comment]
> To what did you compare #1488's RAM usage to??

I compare RAM usage to the master the PR #1488 emerged from.

**Comment by kisslorand** [comment]
After commit https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/1487/commits/314bc35000c07534bb3d4796c1ac848c90796d52 several more bugs are introduced.

Printing from other host:

- If some of the 3 filament statistics info (length, weight, cost) is missing from the printed gCode and previously there was a file printed which had one or more data that is missing from the actual, than after print the missing filament statistics info is still shown but with the data from the previous print.
- If previously printed from TFT and than from another host, in the printing menu the previously printed file name is shown
- printing only from another host than the TFT shows a garbage filename in the menu title during and after print



When printing from TFT, after the print, if you press the "Print finished" area, the popup shows the short (8.3) name of the file printed.



After the print is finished no more print statistics are present in a form of scrollbox.

The "Brief" button brings up a popup that has no filename but only the storage media where the file was printed from.

**Comment by kisslorand** [comment]
I really hoped that's the one but even more bugs are introduced.
 I am tired explaining them so here are some pictures:




This was a quick, 3 minutes test.
There are more bugs but I have to stop, people will say I have something against you.
I see you are getting inspiration from a working and bugless PR (#1488), but this is open source, isn't it?
Not to worry, leave it as it is, I will fix the bugs when/if this PR is merged.

**Comment by rightawy** [comment]
wow, did I miss something? most of the bugs @kisslorand mentioned were present in #1488 if I go through commit by commit and even before this PR. However it is interesting to see how #1488 was rewritten line to line from this one and @kisslorand and now Kisslorand claims his PR inspired guruathwal and trying to defame him. Dude its open source both ways. Its a good thing to report bugs and kisslorand did a good job at this but, this also seems like a seller pointing out defects in other's product to make his look better. But I am no one to judge. The internet is a free world anyone can say anything. I thank all the contributors to this project for improving it. Even if they know programming or not like @oldman4U by doing the maintenance work that no one would do for free and what he getting in return is just pathetic to read. Why whould I say so? Because I am a user of this project. And the least I can do is respect the contributors no matter how much contribute.

**Comment by kisslorand** [comment]
>However it is interesting to see how #1488 was rewritten line to line from this one and @kisslorand now claims his PR inspired guruathwal and trying to defame him.

It's more interesting that there is a thing called "timestamp".
I started to work on the long filename for onboard SD. When I presented PR #1488 it had no bugs. @guruathwal warned me that the way the PR #1488 was, it added too much RAM usage. In my response I described how I started working on this long filename issue and what problems I did encounter. Check here: `https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/1488#discussion_r553891001`. 
I explained there ". I also first started to use only an index of the current file and use it to get the long filename based on that index (from infoFile.Longfile) but that way I couldn't store/get the filename if printing from other USB host (ex Octoprint) that uses onboard SD."
Guess what? 3 or 4 hours later @guruathwal presented a commit that used file index to retrieve the long filename from `infoFile.Longfilename`, a commit that faced the same issue of not being able to retrieve the filename when printing from other source. Later on that day I came up with the idea of storing the value of 65535 in the index variable if the file is not printed from TFT. That's when I returned back to my original idea of storing only the index of the file. 
In the process I moved out the filament statistics info initialization into it's own function and called that function before the printed file is opened. A day later, AFTER I pointed out some bugs about the filament statistics @guruathwal moved out those initializations into a new function and also moved it to be called before the printed file is opened.
I used from the beginning `parseACK` to detect from where is the file opened, TFT or other host. It is done when "File opened:" is received from Marlin. Guess who started to use the same in his latest commit?
The list can go on but it's not the point here. I've been blatantly accused being a fraud and copy "line to line" from this PR #1487.
Someone didn't do his homework properly and forgot about timestamp, being first in the timeline means being the one with the original idea. Cannot copy something that comes later on in the timeline. At least not yet, mindreading is something I cannot do.

Whatever, I still respect @guruathwal's work at least for the simple fact that it gives me great joy learning from it. Optimization is my thing also so I do respect how he is dedicated to that.

**Comment by digant73** [comment]
@guruathwal just to avoid to open a PR for a minor bug introduced by one of the last merged PR, please include the attached ProbeOffsetControl.c. In one of your last PR it was wrongly duplicated a call to "probeHeightStop();      // raise nozzle"

[ProbeOffsetControl.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/5795504/ProbeOffsetControl.zip)

**Comment by rightawy** [comment]
@kisslorand yes I know the timestamps. If you look at the image below this text, you will see a screenshot.  As soon as guruathwal uploaded the code, you copied the changes from it. after this, you briskly changed your opinion from opposing guruathwal's claim to thanking him.
image

One cannot own an idea until he can prove it. Newton did not just have an idea about gravity, he proved it.  Maybe I had this idea even before you did.

**Comment by guruathwal** [comment]
It does not matter to me how other people think of me as a programmer. I myself claim that I am a really bad as a programmer. maybe I will never learn to be as good as programmers. I do not intend to take any credit for anything I did not write. But due to work reasons, since the last couple of months, I am not able to spare enough time for discussions in this repository and access to my machine is also limited. Hence most of my work is based on calculations. 
when I saw @kisslorand's  #1488, I did not believe that it was not possible to reduce the RAM usage. And instead of starting a long discussion I just proved my concept. I was getting mixed in the bug reports by Kisslorand as I had no access to my machine to test the code and with theoretical calculations, it is hard to know if a bug really exists or not.  But my friend @kisslorand, this also confused me when you insisted that using 150 Bytes of RAM was ok in #1488 which contradicted your views about RAM usage (https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/1359#issuecomment-751525484). All I wanted was to prove my concept. I have always tried to keep things as simple as possible for everyone to understand, this also reflects in my work in this repository.
I never intended to claim ownership of any person's work. You say the word and I will remove it from this PR. I do not want any criticism. I always try to avoid conflicts and nonsense debates by not replying to them. But this is looking way over the limits. @kisslorand I believe you that your intention is good and thank you for reporting the bugs in the code. This community have people from many backgrounds and cultures all have different values and ethics maybe I did not get offended by your response but who knows who will. @rightawy I appreciate your support but its not good to stretch it over the limits. My 2 cents: it does not matter what people say, it is how they say. We need to save the dolphins.

Related to **"long file name display during printing from onboard SD"**, here is the video proof that my iteration works with all the scenarios I could imagine:
https://app.box.com/s/no3pxmv3xmm9m1zf670t00o7ushekxyy

**Comment by kisslorand** [comment]
@guruathwal 

In all honesty possible I have to admit this is the most mature post I ever read on this repository! No jokes, no irony, no hidden nothing. This just made me respect the man, not the programmer. Damn! Now I really feel small! Do you know what made me truly believe you post is honest? This:
> But my friend @kisslorand, this also confused me when you insisted that using 150 Bytes of RAM was ok in #1488 which contradicted your views about RAM usage ([#1359 (comment)](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/1359#issuecomment-751525484)).

It tortured me that I couldn't find a full solution without touching the RAM. I spent more than 10 hours to find a way.  I changed countless times that infamous variable's size from 30 to 256. Your comment that it is possible without RAM sacrifice just made the click in my mind how I can do it, by setting the index file to 65535 if the print is from other host.  I was away from any computer the whole day so only later at night I could make that happen.

OK, I promised I will not disclose any more bugs and I wish to keep that promise. I would really like to discuss the ones that remained in private with you. In case you agree, I have no idea how can we share contact details out of public sight.

Respect! (hush @rightawy, I do mean it!)

**Comment by digant73** [comment]
@guruatwhal not yet but remember that my PRs removing Zraise and introducing xy probing config param fixed the bugs related to nozzle raising and scratching the bed (on some particular printer setup). For that reason z raise was made configurable. In this moment it can be executed even twice. I think you mixed the old and new code when you changed the code style with your PR. Just an error  during a git resolve

**Comment by digant73** [comment]
@guruatwhal not yet but remember that my PRs removing Zraise and introducing xy probing config param fixed the bugs related to nozzle raising and scratching the bed (on some particular printer setup). For that reason z raise was made configurable. In this moment it can be executed even twice. I think you mixed the old and new code when you changed the code style with your PR. Just an error  during a git resolve

**Comment by kisslorand** [comment]
>I never intended to claim ownership of any person's work. You say the word and I will remove it from this PR. I do not want any criticism. 

Since you're asking for it, here it is.

https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/1488#discussion_r553891001 :
"use only an index of the current file and use it to get the long filename based on that index (from infoFile.Longfile)"
This is exactly what the long filename display foundation is in this PR.

I do not want to start anything, it is also an explanation why I said that "inspiration" contributed to the solution for the long filename issue when printing from onboard SD.

**Comment by guruathwal** [comment]
> > I never intended to claim ownership of any person's work. You say the word and I will remove it from this PR. I do not want any criticism.
> 
> Since you're asking for it, here it is.
> 
> [#1488 (comment)](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/pull/1488#discussion_r553891001) :
> "use only an index of the current file and use it to get the long filename based on that index (from infoFile.Longfile)"
> This is exactly the foundation for the long filename display in this PR.

How else should it be explained what the PR contains? I use that format for describing the changes in all my PRs. Even when @digant73 suggested changes in the code separately in my earlier PRs, I made the required changes to them and posted the changes in the same style. Do this mean that I claimed ownership of those changes provided by digant73?

> I do not want to start anything, it is also an explanation why I said that "inspiration" was used in the solution for the long filename issue when printing from onboard SD.

I do not think this is going anywhere. I think I made a mistake by replying to these comments. I never claimed that I came up with the idea or concept and I never said that this is my original idea. Why are you trying to force the blame when I specifically said that I proved my point for low ram usage. It is puzzling, why is it so that earlier you rejected you "Idea" saying that it was not possible but when I proved that It was surely possible (I also provided a video [here](https://app.box.com/s/no3pxmv3xmm9m1zf670t00o7ushekxyy)), what are you being so insecure after that? Why did you provided the bug report? was it to suggest that what I proved is false?

What if I had never commented on your PR on first place and provided the changes after it was merged will this situation be same then?

>OK, I promised I will not disclose any more bugs and I wish to keep that promise. I would really like to discuss the ones that remained in private with you. In case you agree, I have no idea how can we share contact details out of public sight.

I never asked you to stop reporting bugs and never asked any sort of promise for the same. I would very much like everything to be public to avoid any future claims and conflicts.

**Comment by rightawy** [comment]
He has an ego problem.

**Comment by bigtreetech** [comment]
@guruathwal Hello. Can this PR be merged now?

**Comment by guruathwal** [comment]
yes, please merge it.

**Pull request merged**